src/main/java/org/spongepowered/common/block/SpongeTileEntityArchetypeBuilder.java

```diff
@@ -0,0 +1,197 @@
+        final TileEntity tileEntity = location.getTileEntity()
+                .orElseThrow(() -> new IllegalArgumentException("There is no tile entity available at the provided location: " + location));
+
+        return tile(tileEntity);
+    }
+
+    @Override
+    public TileEntityArchetype.Builder tile(TileEntity tileEntity) {
+        checkArgument(tileEntity instanceof net.minecraft.tileentity.TileEntity, "TileEntity is not compatible with this implementation!");
+        NBTTagCompound nbttagcompound = new NBTTagCompound();
+        ((net.minecraft.tileentity.TileEntity) tileEntity).writeToNBT(nbttagcompound);
+        nbttagcompound.removeTag("x");
+        nbttagcompound.removeTag("y");
+        nbttagcompound.removeTag("z");
+        this.tileData = NbtTranslator.getInstance().translate(nbttagcompound);
+        this.blockState = tileEntity.getBlock();
+        this.tileEntityType = tileEntity.getType();
+        return this;
+    }
+
+    @Override
+    public TileEntityArchetype.Builder tileData(DataView dataView) {
+        checkNotNull(dataView, "Provided DataView cannot be null!");
+        final DataContainer copy = dataView.copy();
+        SpongeDataManager.getInstance().getValidators(Validations.TILE_ENTITY).validate(copy);
+        this.tileData = copy;
+        return this;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public TileEntityArchetype.Builder setData(DataManipulator<?, ?> manipulator) {
+        if (this.tileData == null) {
+            this.tileData = new MemoryDataContainer();
+        }
+        SpongeDataManager.getInstance().getRawNbtProcessor(NbtDataTypes.TILE_ENTITY, manipulator.getClass())
+                .ifPresent(processor -> processor.storeToView(this.tileData, manipulator));
+        return this;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <E, V extends BaseValue<E>> TileEntityArchetype.Builder set(V value) {
+        if (this.tileData == null) {
+            this.tileData = new MemoryDataContainer();
+        }
+        SpongeDataManager.getInstance().getRawNbtProcessor(NbtDataTypes.TILE_ENTITY, value.getKey())
+                .ifPresent(processor -> processor.offer(this.tileData, value));
+        return this;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public <E, V extends BaseValue<E>> TileEntityArchetype.Builder set(Key<V> key, E value) {
+        if (this.tileData == null) {
+            this.tileData = new MemoryDataContainer();
+        }
+        SpongeDataManager.getInstance().getRawNbtProcessor(NbtDataTypes.TILE_ENTITY, key)
+                .ifPresent(processor -> processor.offer(this.tileData, value));
+        return this;
+    }
+
+    @Override
+    public TileEntityArchetype build() {
+        checkState(this.blockState != null, "BlockState cannot be null!");
+        checkState(this.tileEntityType != null, "TileEntityType cannot be null!");
+        if (this.tileData != null) {
```

Comment:
I do not really understand why you are performing this check...
Because the problem is that this means the build Archetype will not contain the source block's tile data (i.e. a Sign won't keep its text).

So is this a choice or a mistake ? I think this is related to what @ZephireNZ said : 

> When converting from Entity to EntityArchetype all data including the ID is in archetype's data. However when built using a builder it is not as far as I can tell.